### PR TITLE
Make PlayerArmorChangeEvent not fire on player join

### PR DESCRIPTION
**Pull request by Silvre**
As the name of the event suggests, it seems unfitting that the event is called when the player joins a server.  I can't think of a use case where you would want this event to call when a player joins, and that could easily be handled on PlayerJoinEvent or similar.

I also updated the description of the event in the javadocs to more accurately reflect when the event fires.

~~An afterthought I had while writing this is that the event be renamed PlayerArmorUpdateEvent.~~

### CODE REVIEW

File: patches/api/0072-Add-PlayerArmorChangeEvent.patch [hunk 1]
```diff
@@ -28,9 +28,9 @@ index 0000000000000000000000000000000000000000..e406ce639a2e88b78f82f25e71678a66
 +import static org.bukkit.Material.*;
 +
 +/**
-+ * Called when the player themselves change their armor items
++ * Called when the player's armor is changed(manual armor change, equip through dispenser, loss of durability, etc.)
 + * <p>
-+ * Not currently called for environmental factors though it <strong>MAY BE IN THE FUTURE</strong>
++ * This event does <strong>not</strong> fire when a player joins the server with armor equipped.
```
Comment (Machine-Maker) [inline patches/api/0072-Add-PlayerArmorChangeEvent.patch]: Is this clarification needed? I wouldn't expect this to fire there anyways, so I'm not sure what this is clarifying.

**Review by Machine-Maker — CHANGES_REQUESTED** [review]
I think (hopefully) the event firing on player join can be considered a bug. Craftbukkit added, [in a fix](https://hub.spigotmc.org/stash/projects/SPIGOT/repos/craftbukkit/commits/c3e684ef704c2685bc85b1f2b760d7075e434dfc#nms-patches/EntityTrackerEntry.patch) so I imagine the event was never intended to be called on player join.

File: patches/api/0072-Add-PlayerArmorChangeEvent.patch (hunk 1 above)
Comment (qixils) [reply patches/api/0072-Add-PlayerArmorChangeEvent.patch]: it's likely there to document the change in functionality, though as this behavior was undocumented i _think_ it's fine just to leave out

**Comment by lexikiq** [comment]
> An afterthought I had while writing this is that the event be renamed PlayerArmorUpdateEvent.
i don't think that'd be ideal-- it means basically the same thing and would only serve to break plugins :P some relevant discussion on that front can be found in #5893

**Review by qixils — COMMENTED** [review]
fine with the server patch, just confused about the javadoc changes

> An afterthought I had while writing this is that the event be renamed PlayerArmorUpdateEvent.

i don't think that'd be ideal-- it means basically the same thing and would only serve to break plugins :P some relevant discussion on that front can be found in #5893

File: patches/api/0072-Add-PlayerArmorChangeEvent.patch (hunk 1 above)
Comment (Silvre) [reply patches/api/0072-Add-PlayerArmorChangeEvent.patch]: I was on the fence about putting it in, on the _very small_ edge case that someone use it on player join, but that is a good point.

**Review by Machine-Maker — APPROVED** [review]

**Comment by qixils** [comment]
i swear i suggested some changes but they seem to have disappeared... thanks github. anyway, there's a missing parenthesis in the javadocs: `changed(manual`, though tbh i'd suggest rewording it to avoid parentheses entirely, something like: `Called when the player's armor is changed via the player equipping it, a dispenser equipping it onto the player, the armor taking durability damage, etc.`

**Comment by Crypnotic** [comment]
Any update on this? Causes issues when listening to armor equipping to modify persistent data values.

File: patches/server/0164-Add-PlayerArmorChangeEvent.patch [hunk 2]
```diff
@@ -20,7 +20,7 @@ index 8044082ed3ca6076af38e4299e50f1f690d02a72..0dc03f53c80aac91a914bea91958a6c1
  
              if (!ItemStack.matches(itemstack1, itemstack)) {
 +                // Paper start - PlayerArmorChangeEvent
-+                if (this instanceof ServerPlayer && enumitemslot.getType() == EquipmentSlot.Type.ARMOR) {
++                if (this instanceof ServerPlayer && enumitemslot.getType() == EquipmentSlot.Type.ARMOR && !((ServerPlayer) this).joining) {
```
Comment (Machine-Maker) [inline patches/server/0164-Add-PlayerArmorChangeEvent.patch:23]: This probably isn't the right solution. It looks like there are other instances where collectEquipmentChanges could be called. Once place I see, is spectating an entity that isn't tracking the player, I couldn't think of when that might happen, but the way the code is written, that would *also* fire this event.

**Review by Machine-Maker — CHANGES_REQUESTED** [review]
Need to figure out if there is any other situation where the event is fired.

File: patches/server/0164-Add-PlayerArmorChangeEvent.patch (hunk 2 above)
Comment (Silvre) [reply patches/server/0164-Add-PlayerArmorChangeEvent.patch:23]: Honestly forgot I made this PR haha, but yeah;  after initially figuring out how the event works it is inherently flawed.  I only planned for this PR to fix when this event occurs on player join, but it might be a good idea to have some discussion for another PR that defines exactly when this event should fire.

Comment (Machine-Maker) [reply patches/server/0164-Add-PlayerArmorChangeEvent.patch:23]: So I can think of two possible solutions, well one fully defined one, one more complicated.
1. Add a boolean parameter to the 2 methods in LivingEntity, `detectEquipmentChanges`, and `collectEquipmentChanges` to determine if the event should be called.
2. Find a better way fix to the issue spigot fixed which caused this event to fire in the wrong spot.

Comment (Machine-Maker) [reply patches/server/0164-Add-PlayerArmorChangeEvent.patch:23]: I think this is actually the right place for this now. So I think this is the correct fix.

**Review by Machine-Maker — APPROVED** [review]

**Review by electronicboy — APPROVED** [review]

**Review by Owen1212055 — CHANGES_REQUESTED** [review]
After testing this fix, this doesn't seem to work anymore as of Paper 1.19.2.

It seems like entities are ticked first, THEN their connection is.
```java
[16:33:59 INFO]: [STDOUT] [net.minecraft.server.level.ServerPlayer] No longer joining
[16:33:59 WARN]: java.lang.NullPointerException - // setting joining to false
[16:33:59 WARN]: 	at net.minecraft.server.level.ServerPlayer.tick(ServerPlayer.java:636)
[16:33:59 WARN]: 	at net.minecraft.server.level.ServerLevel.tickNonPassenger(ServerLevel.java:1192)
[16:33:59 WARN]: 	at net.minecraft.world.level.Level.guardEntityTick(Level.java:897)
[16:33:59 WARN]: 	at net.minecraft.server.level.ServerLevel.lambda$tick$6(ServerLevel.java:730)
[16:33:59 WARN]: 	at net.minecraft.world.level.entity.EntityTickList.forEach(EntityTickList.java:42)
[16:33:59 WARN]: 	at net.minecraft.server.level.ServerLevel.tick(ServerLevel.java:710)
[16:33:59 WARN]: 	at net.minecraft.server.MinecraftServer.tickChildren(MinecraftServer.java:1535)
[16:33:59 WARN]: 	at net.minecraft.server.dedicated.DedicatedServer.tickChildren(DedicatedServer.java:446)
[16:33:59 WARN]: 	at net.minecraft.server.MinecraftServer.tickServer(MinecraftServer.java:1397)
[16:33:59 WARN]: 	at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1173)
[16:33:59 WARN]: 	at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:305)
[16:33:59 WARN]: 	at java.base/java.lang.Thread.run(Thread.java:833)
[16:33:59 WARN]: java.lang.NullPointerException - // ArmorChangeEvent
[16:33:59 WARN]: 	at test-plugin-1.0.0-SNAPSHOT.jar//io.papermc.paper.testplugin.TestPlugin.testEvent(TestPlugin.java:19)
[16:33:59 WARN]: 	at com.destroystokyo.paper.event.executor.asm.generated.GeneratedEventExecutor1.execute(Unknown Source)
[16:33:59 WARN]: 	at org.bukkit.plugin.EventExecutor.lambda$create$1(EventExecutor.java:75)
[16:33:59 WARN]: 	at co.aikar.timings.TimedEventExecutor.execute(TimedEventExecutor.java:80)
[16:33:59 WARN]: 	at org.bukkit.plugin.RegisteredListener.callEvent(RegisteredListener.java:70)
[16:33:59 WARN]: 	at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:670)
[16:33:59 WARN]: 	at org.bukkit.event.Event.callEvent(Event.java:45)
[16:33:59 WARN]: 	at net.minecraft.world.entity.LivingEntity.collectEquipmentChanges(LivingEntity.java:3077)
[16:33:59 WARN]: 	at net.minecraft.world.entity.LivingEntity.detectEquipmentUpdates(LivingEntity.java:3038)
[16:33:59 WARN]: 	at net.minecraft.world.entity.LivingEntity.tick(LivingEntity.java:2964)
[16:33:59 WARN]: 	at net.minecraft.world.entity.player.Player.tick(Player.java:277)
[16:33:59 WARN]: 	at net.minecraft.server.level.ServerPlayer.doTick(ServerPlayer.java:683)
[16:33:59 WARN]: 	at net.minecraft.server.network.ServerGamePacketListenerImpl.tick(ServerGamePacketListenerImpl.java:365)
[16:33:59 WARN]: 	at net.minecraft.network.Connection.tick(Connection.java:579)
[16:33:59 WARN]: 	at net.minecraft.server.network.ServerConnectionListener.tick(ServerConnectionListener.java:233)
[16:33:59 WARN]: 	at net.minecraft.server.MinecraftServer.tickChildren(MinecraftServer.java:1565)
[16:33:59 WARN]: 	at net.minecraft.server.dedicated.DedicatedServer.tickChildren(DedicatedServer.java:446)
[16:33:59 WARN]: 	at net.minecraft.server.MinecraftServer.tickServer(MinecraftServer.java:1397)
[16:33:59 WARN]: 	at net.minecraft.server.MinecraftServer.runServer(MinecraftServer.java:1173)
[16:33:59 WARN]: 	at net.minecraft.server.MinecraftServer.lambda$spin$0(MinecraftServer.java:305)
[16:33:59 WARN]: 	at java.base/java.lang.Thread.run(Thread.java:833)
```

**Comment by Owen1212055** [comment]
Because it seems like this fix no longer applies, this will be closed, leave a comment and this can be reopened.
If someone is interested in finding another fix for this, feel free to open another PR!
😄
